ddtrace/internal/telemetry/writer.py

```diff
@@ -158,8 +184,49 @@ def _stop_service(self, *args, **kwargs):
         super(TelemetryWriter, self)._stop_service(*args, **kwargs)
         self.join()
 
+    def add_gauge_metric(self, namespace, name, value, tags={}):
+        # type: (str,str, float, MetricTagType) -> None
+        """
+        Queues count metric
+        """
+        self._add_metric("gauge", namespace, name, value, tags)
+
+    def add_rate_metric(self, namespace, name, value=1.0, tags={}):
+        # type: (str,str, float, MetricTagType) -> None
+        """
+        Queues count metric
+        """
+        self._add_metric("rate", namespace, name, value, tags)
+
+    def add_count_metric(self, namespace, name, value=1.0, tags={}):
+        # type: (str,str, float, MetricTagType) -> None
+        """
+        Queues count metric
+        """
+        self._add_metric("count", namespace, name, value, tags)
+
+    def _add_metric(self, metric_type, namespace, name, value=1.0, tags={}):
+        # type: (MetricType, str,str, float, MetricTagType) -> None
+        """
+        Queues metric
+        """
+        with self._lock:
```

Comment:
This lock scares me. I can see a scenario where we can inadvertently introduce deadlocks if decide to collect telemetry metrics in certain parts of our codebase. Not sure what we can do here. Just wanted to call it out.